libsignal-protocol/src/crypto/native.rs

```diff
@@ -1,17 +1,112 @@
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+    ) -> Result<Vec<u8>, InternalError> {
+        let result = match (cipher, key.len()) {
+            (SignalCipherType::AesCtrNoPadding, 16) => {
+                let mut buf = data.to_vec();
+                // a side note here is that Ctr mode takes a nonce as 2nd param,
+                // not an IV but if we for example use a
+                // randomly nonce generated every call, it would fail
+                // i mean by fail here, fail in decryption the cipher text
+                // created by the same key. so i think it's not
+                // not meant to be nonce here, but just an IV.
+                // and now it works same as the openssl
+                let mut c = Aes128Ctr::new_var(key, iv)
+                    .map_err(|_| InternalError::Unknown)?;
+                c.apply_keystream(&mut buf);
+                buf
+            },
+            (SignalCipherType::AesCtrNoPadding, 24) => {
+                let mut buf = data.to_vec();
+                let mut c = Aes192Ctr::new_var(key, iv)
+                    .map_err(|_| InternalError::Unknown)?;
+                c.apply_keystream(&mut buf);
+                buf
+            },
+            (SignalCipherType::AesCtrNoPadding, 32) => {
+                let mut buf = data.to_vec();
+                let mut c = Aes256Ctr::new_var(key, iv)
+                    .map_err(|_| InternalError::Unknown)?;
+                c.apply_keystream(&mut buf);
+                buf
+            },
+            (SignalCipherType::AesCbcPkcs5, 16) => {
+                let c = Aes128Cbc::new_var(&key, &iv)
+                    .map_err(|_| InternalError::Unknown)?;
+                let buf = match mode {
+                    Mode::Encrypt => c.encrypt_vec(data),
+                    Mode::Decrypt => c
+                        .decrypt_vec(data)
+                        .map_err(|_| InternalError::Unknown)?,
+                };
+                buf
+            },
+            (SignalCipherType::AesCbcPkcs5, 24) => {
+                let c = Aes192Cbc::new_var(&key, &iv)
+                    .map_err(|_| InternalError::Unknown)?;
+                let buf = match mode {
+                    Mode::Encrypt => c.encrypt_vec(data),
+                    Mode::Decrypt => c
+                        .decrypt_vec(data)
+                        .map_err(|_| InternalError::Unknown)?,
+                };
+                buf
+            },
+            (SignalCipherType::AesCbcPkcs5, 32) => {
+                let c = Aes256Cbc::new_var(&key, &iv)
+                    .map_err(|_| InternalError::Unknown)?;
+                let buf = match mode {
+                    Mode::Encrypt => c.encrypt_vec(data),
+                    Mode::Decrypt => c
+                        .decrypt_vec(data)
+                        .map_err(|_| InternalError::Unknown)?,
+                };
+                buf
+            },
+            _ => unreachable!(),
```

Comment:
We probably want to add more description to this `unreachable!()`. For example, `(cipher, size) => unreachable!("A combination of {} and {} doesn't make sense", cipher, size)`.